lms/djangoapps/learning_success/management/commands/challenges_helper.py

```diff
@@ -34,68 +52,198 @@ def generate_default_skills(skill_tags):
-                            in challenge_counter.keys()}
-    student_skills = deepcopy(default_skills)
-
-    for submission in student.challengesubmission_set.all():
-        module = challenge_index[submission.challenge_id]
-        challenge_tags = skill_tags[submission.challenge_id]
-        if submission.passed:
-            challenge_activities[module]['passed'] += 1
-            increment_student_skill_tags(student_skills, challenge_tags)
-        else:
-            challenge_activities[module]['attempted'] += 1
-        challenge_activities[module]['num_attempts'] += submission.attempts
-
-    for module_level, total_challenges in challenge_counter.items():
-        activities = challenge_activities[module_level]
-        activities['unattempted'] = (
-            total_challenges - activities['passed'] - activities['attempted'])
-        activities.setdefault('num_attempts', 0)
-
-    challenge_activities = {
-        module: json.dumps(challenges)
-        for module, challenges in challenge_activities.items()
-    }
-    challenge_activities['student_skills'] = json.dumps(student_skills)
-    return challenge_activities
-
-
-def extract_all_student_challenges(program):
-    """ Calculates the historical challenge data for all students
-
-    Returns a dict with email and challenge history for each student """
-    challenge_index, skill_tags = index_challenge_to_module_and_level()
-    default_skills = generate_default_skills(skill_tags)
-    challenge_counter = Counter(challenge_index.values())
-    students = program.enrolled_students.all()
-    challenge_history = {
-        student.email: single_student_challenge_history(
-            student, challenge_counter, challenge_index, skill_tags,
-            default_skills)
-        for student in students.prefetch_related('challengesubmission_set')
+def extract_challenge_submissions():
+    """ Queries the mongodb for all challenge submissions and aggregates the
+    results in such a way that we get the most recent submission's passed
+    status (T/F), number of attempts grouped by challenge and user
+
+    Returns a dict with a tuple of (challenge_id, user_id) as key and
+    the submission dict as value """
+    MONGO_DB = settings.MONGO_CLIENT['challenges']
+    mongo_submissions = MONGO_DB['submissions'].aggregate([
+        { "$sort": {"submitted":-1, "challenge_id": 1, "user_id": 1} },
+        { "$group": {
+                "_id" : "$challenge_id",
+                "submitted": {"$last": "$submitted" },
+                "user_id": {"$first": "$user_id" },
+                "passed": {"$first": "$passed" },
+                "attempts": {"$sum": 1}
+            }
+        }
+    ])
+    return {(str(submission['_id']), submission['user_id']): submission
+            for submission in mongo_submissions}
+
+
+def get_all_published_blocks():
+    """ Queries the mongo database for all published blocks and returns the
+    results as a list of dicts"""
+    query = [
```

Comment:
Just for comparison - have a quick look at this guy and tell me what you think :) https://github.com/Code-Institute-Org/ci_sandbox/blob/main/django_proxy/django_ci_sandbox_proxy/ci_sandbox/xblock.py

Reply:
I think this is really cool :+1: Have not tried it out outside of the sandbox, but would you be in favour of replacing my query functionality with this for better maintenance? We already have a previous version of this in the edx codebase, but I did not want to fiddle with it as I was not sure it that the other functionality would require the same data structure. What do you think?

Reply:
Yeah I'd leave the edx implementation where it is. This one is a handy number for iterating over different levels of blocks. We might need to find a good place to put him, if he's used across multiple projects.